src/libopensc/card-iasecc.c

```diff
@@ -3181,45 +3208,91 @@ iasecc_qsign_data_sha1(struct sc_context *ctx, const unsigned char *in, size_t i
 	       in_len);
 	memset(out, 0, sizeof(struct iasecc_qsign_data));
 
-	SHA1_Init(&sha);
-	SHA1_Update(&sha, in, in_len);
+	md = EVP_get_digestbyname("SHA1");
+	mdctx = EVP_MD_CTX_new();
+        if (EVP_DigestInit_ex(mdctx, md, NULL) != 1) {
+		sc_log(ctx, "EVP_DigestInit_ex failed");
+		goto err;
+	}
+	
+#if defined(LIBRESSL_VERSION_NUMBER)
+	md_data = (SHA_CTX *)mdctx->md_data;
+#elif  OPENSSL_VERSION_NUMBER < 0x30000000L
+	md_data = EVP_MD_CTX_md_data(mdctx);
+#else
+	md_data = EVP_MD_CTX_get0_md_data(mdctx);
+#endif
```

Comment:
I will fix the white space.
In regards to `card-iasecc.c` I know you wanted the defines in `sc-ossl-compat.h`.  But here is my thinking on `card-iasecc.c` and `sc-ossl-compat,h`:

- No one as said they even want to test with the current code the `card-iasecc.c` code. 
- The Specs for this card are from 2008 and appeard to be: https://dvv.fi/documents/16079645/17324992/IAS+ECC+v1_0_1UK.pdf/84043361-56ad-bc26-8223-37e0d79355d8/IAS+ECC+v1_0_1UK.pdf?t=1622729405351
- This may be a unused driver as the IAE-ECC are for gov issued cards.
- `card-iasecc.c`  is the only source code that looks at the `md_data`.
- OpenSSL-3.0.0 developers do not want to expose the `md_data` Today, `md_data = `EVP_MD_CTX_get0_md_data(mdctx);` returns NULL, but they also said `EVP_MD_CTX_get0_md_data` should be deprecated This could happen soon. 
- `md_data = EVP_MD_CTX_md_data(mdctx);` compiles with OpenSSL 1.1.1
- `md_data = (SHA_CTX *)mdctx->md_data;`  compiles with LibreSSL
- If we do option (2.) "Include an internal SHA256 function used only for cards doing this last round of hash on card." it should be used for these two routines and not use LibreSSL or OpeenSSL-1.1.1  so no \#ifdefs would be needed. 
- The code is already hard to read, you need to understand what is in `SHA_CTX`, `SHA256_CTX` and  the specs listed above in  "6.1.1.3 Protocol steps" and "9.6 Perform security operation commands" that describes the data sent and the APDU. 

If I was going to change `card-iasecc.c` I would put `#if OPENSSL_VERSION_NUMBER >= 0x30000000L`  in `iasecc_qsign_data_sha1` and `iasecc_qsign_data_sha256` and  return SC_ERROR_NOT_SUPPORTED 

And since `card-iasecc.c` appears to be the last issue for supporting OpenSSL-3.0.0, I did not want to pollute `sc-ossl-compat.h` with defines we will most like drop if we drop the driver or add an OpenSC SHA256.

Reply:
See https://github.com/OpenSC/OpenSC/pull/2506/commits/a9234ee8fdb7e0e0f872de7fd4a793fa86a79f73  which will run as before, but when compiled with OpenSSL 3.0 returns SC_ERROR_NOT_SUPPORTED.